### PR TITLE
fix(modeling_ppo): load reference head under zero3

### PR DESCRIPTION
**Pull request by maxreciprocate**
Fix of #461 #483

Under zero3 the reference head will be kept unsharded. I couldn't get sharding to work because deepspeed expects sharded parameters to have at least one backward pass, while there is no backward pass for the reference head.

Check:
```bash
accelerate launch --config_file configs/accelerate/zero3.yaml examples/ppo_sentiments.py
```

https://wandb.ai/sorry/trlx-references/reports/fix-ppo-zero3-v-main--Vmlldzo0NjY3MDYy

### CODE REVIEW

**Comment by nikebless** [comment]
I don't think unsharding is necessary for zero3 in general. It is only necessary for `zero3_init_flag: true`.

I ran your changes locally, undoing the `modeling_ppo.py` changes and updating `configs/accelerate/zero3.yaml` to `zero3_init_flag: false`, and the training worked (side note: passing `--zero3_init_flag false` as an argument doesn't override the config file)

File: trlx/models/modeling_ppo.py [hunk 1]
```diff
@@ -410,10 +410,18 @@ def __init__(
...
-        decoder_blocks = deepcopy(hf_get_decoder_blocks(base_model))
-        self.decoder_blocks = nn.ModuleList(list(decoder_blocks)[-num_layers_unfrozen:])
-        self.final_norm = deepcopy(hf_get_decoder_final_norm(base_model))
-        self.lm_head = deepcopy(hf_get_lm_head(base_model))
+
+        decoder_blocks = hf_get_decoder_blocks(base_model)[-num_layers_unfrozen:]
+        final_norm = hf_get_decoder_final_norm(base_model)
+        lm_head = hf_get_lm_head(base_model)
+
+        with deepspeed.zero.GatheredParameters(
```
Comment (Dahoas) [inline trlx/models/modeling_ppo.py:418]: Should this be dependent on whether we are using zero3 (zero3_init)?

Comment (Dahoas) [reply trlx/models/modeling_ppo.py:418]: @maxreciprocate

Comment (maxreciprocate) [reply trlx/models/modeling_ppo.py:418]: @Dahoas Not necessary, this context would be a noop when zero3 is not enabled and would still work correctly, see comparisons runs https://wandb.ai/sorry/trlx-references/reports/fix-ppo-zero3-v-main--Vmlldzo0NjY3MDYy

**Review by Dahoas — APPROVED** [review]
Nice, if you have no other changes let's merge this in.

**Pull request merged**